lib/Charts/BaseChart.js

```diff
@@ -0,0 +1,392 @@
+
+        const svg = d3Element.append('svg')
+            .attr('class', 'd3')
+            .attr('width', state.width)
+            .attr('height', state.height);
+
+        const chart = svg.append('g')
+            .attr('class', 'line-chart');
+
+        chart.append('rect')
+            .attr('class', 'plot-area')
+            .attr('x', '0')
+            .attr('y', '0');
+        chart.append('g')
+            .attr('class', 'x axis')
+            .style('opacity', 1e-6)
+            .append('text')
+            .attr('class', 'label');
+        chart.append('g')
+            .attr('class', 'y axes');
+        chart.append('g')
+            .attr('class', 'no-data')
+            .append('text');
+        chart.append('g')
+            .attr('class', 'selection');
+
+        Tooltip.create(state.tooltipSettings);
+
+        this.update(element, state);
+    },
+
+    update(element, state) {
+        this.render(element, state);
+    },
+
+    destroy(element, state) {
+        Tooltip.destroy(state.tooltipSettings);
+    },
+    
+    render(element, state) {
+        if (!defined(state.data)) {
+            return;
+        }
+        const data = state.data;
+        
+        const margin = defaultValue(state.margin, defaultMargin);
+        const allUnits = getUniqueValues(state.data.map(line => line.units));
+        const size = Size.calculate(element, margin, state, state.mini ? 1 : allUnits.length);
+        const scales = Scales.calculate(size, state.domain, state.data, this.getXpadding(data));
+        
+        const transitionDuration = defaultValue(state.transitionDuration, defaultTransitionDuration);
+        const hasData = (data.length > 0 && data[0].points.length > 0);
+
+        const d3Element = d3Select(element);
+        
+        // Update SVG
+        const svg = d3Element.select('svg')
+            .attr('width', state.width)
+            .attr('height', state.height)
+            .style('user-select', 'none')
+            .style('cursor', 'default');
+
+        // Update Plot Area
+        const plotArea = d3Element.select('rect.plot-area');
+        plotArea.attr('width', size.width)
+            .attr('height', size.plotHeight);
+
+        // Update chart area
+        const gTransform = 'translate(' + (margin.left + size.yAxesWidth) + ',' + (margin.top + Title.getHeight(state.titleSettings)) + ')';
+        const g = d3Select(element).selectAll('.line-chart').attr('transform', gTransform);
+        const t = d3Transition().duration(transitionDuration);
+
+        // Title.
+        Title.enterUpdateAndExit(d3Element, state.titleSettings, margin, data, transitionDuration);
+
+        const renderConext = {
+            margin,
+            allUnits,
+            size,
+            scales,
+            data,
+            transitionDuration,
+            hasData,
+            d3Element,
+            svg,
+            gTransform,
+            g,
+            t,
+            state,
+            element
+        };
```

Comment:
This is quite a funny old bunch of things to pass around like this. It would be much nicer to figure out which ones we really need and put them into a defined class.